README
NAME
    Dist::Zilla::Plugin::Prereqs::MatchInstalled - Depend on versions of
    modules the same as you have installed

VERSION
    version 1.001003

SYNOPSIS
    This is based on the code of
    "[Author::KENTNL::Prereqs::Latest::Selective]", but intended for a wider
    audience.

        [Prereqs::MatchInstalled]
        module = My::Module

    If you want to automatically add all modules that are "prereqs", perhaps
    instead look at "[Prereqs::MatchInstalled::All]"

    NOTE: Dependencies will only be upgraded to match the *Installed*
    version if they're found elsewhere in the dependency tree.

    This is designed so that it integrates with other automated version
    provisioning.

    If you're hard-coding module dependencies instead, you will want to
    place this module *after* other modules that declare dependencies.

    For instance:

        [Prereqs]
        Foo = 0

        [Prereqs::MatchInstalled]
        module = Foo

    ^^ "Foo" will be upgraded to the version installed.

    By default, dependencies that match values of "module" will be upgraded
    when they are found in:

        phase: build, test, runtime, configure, develop
        relation: depends, suggests, recommends

    To change this behavior, specify one or more of the following
    parameters:

        applyto_phase = build
        applyto_phase = configure

        applyto_relation = requires

    etc.

    For more complex demands, this also works:

        applyto = build.requires
        applyto = configure.recommends

    And that should hopefully be sufficient to cover any conceivable
    use-case.

    Also note, we don't do any sort of sanity checking on the module list
    you provide.

    For instance,

        module = strict
        module = warning

    Will both upgrade the strict and warnings dependencies on your module,
    regardless of how daft an idea that may be.

    And with a little glue

        module = perl

    Does what you want, but you probably shouldn't rely on that :).

METHODS
  mvp_multivalue_args
    The following properties can be specified multiple times:

    *   "applyto"

    *   "applyto_relation"

    *   "applyto_phase"

    *   "modules"

  "mvp_aliases"
    The "module" is an alias for "modules"

  "current_version_of"
        $self->current_version_of($package);

    Attempts to find the current version of $package.

    Returns "undef" if something went wrong.

  "register_prereqs"
    This is for "Dist::Zilla::Role::PrereqSource", which gets new
    prerequisites from this module.

ATTRIBUTES
  "applyto_phase"
    Determines which phases will be checked for module dependencies to
    upgrade.

        [Prereqs::MatchInstalled]
        applyto_phase = build
        applyto_phase = test

    Defaults to:

        build test runtime configure develop

  "applyto_relation"
    Determines which relations will be checked for module dependencies to
    upgrade.

        [Prereqs::MatchInstalled]
        applyto_relation = requires

    Defaults to:

        requires suggests recommends

  "applyto"
    Determines the total list of "phase"/"relation" combinations which will
    be checked for dependencies to upgrade.

    If not specified, is built from "applyto_phase" and "applyto_relation"

        [Prereqs::MatchInstalled]
        applyto = runtime.requires
        applyto = configure.requires

  "modules"
    Contains the list of modules that will be searched for in the existing
    "Prereqs" stash to upgrade.

        [Prereqs::MatchInstalled]
        module = Foo
        module = Bar
        modules = Baz ; this is the same as the previous 2

    If you want to automatically add all modules that are "prereqs", perhaps
    instead look at "[Prereqs::MatchInstalled::All]"

PRIVATE ATTRIBUTES
  "_applyto_list"
    Internal.

    Contains the contents of "applyto" represented as an
    "ArrayRef[ArrayRef[Str]]"

  "_modules_hash"
    Contains a copy of "modules" as a hash for easy look-up.

PRIVATE METHODS
  _build_applyto
  _build_applyto_list
  _build__modules_hash
  _user_wants_upgrade_on
AUTHOR
    Kent Fredric <kentnl@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for Dist-Zilla-Plugin-Prereqs-MatchInstalled

1.001004 2017-03-03T08:41:27Z 1da8b6c

1.001003 2017-03-03T08:16:12Z 86150e7
 [Bugfix]
 - Avoid test failures on 5.26 w/ -Ddefault_inc_excludes_dot
 - Closes https://github.com/kentnl/Dist-Zilla-Plugin-Prereqs-MatchInstalled/issues/4

 [Dependencies::Stats]
 - Dependencies changed since 1.001002, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +14 ↑4 -3 (recommends: +1, suggests: ↑2)
 - runtime: (recommends: +1)
 - test: +1 ↓1 (recommends: ↑1 ↓1)

1.001002 2014-09-07T08:54:48Z
 [Bugfix]
 - Don't fail on tests if Test::Difference is not installed.
 - Closes #3
 - Thanks Sander Lepik for reporting.

 [Dependencies::Stats]
 - Dependencies changed since 1.001001, see misc/*.deps* for details
 - develop: (suggests: ↑1)

1.001001 2014-09-05T03:26:01Z
 [Bugfix]
 - Don't be lazy, get test data comparison correctly.

 [Dependencies::Stats]
 - Dependencies changed since 1.001000, see misc/*.deps* for details
 - test: +1

1.001000 2014-09-04T21:06:03Z
 [00 Minor]
 - Internals changed.
 - Tests refactored and improved.
 - metadata

 [Dependencies::Stats]
 - Dependencies changed since 1.000002, see misc/*.deps* for details
 - develop: +1 ↑1 (suggests: ↑2)
 - runtime: +1
 - test: +1 -3 (recommends: ↑1)

 [Internals]
 - dump_config implemented in config_dumper

 [Metadata]
 - set author = kentnl

 [Tests]
 - rewritten in dztest and enhanced slightly.

1.000002 2014-08-15T21:18:20Z
 [00 Trivial]
 - CPANDAY!
 - no code changes.
 - tests augmented.
 - whitespace adjusted.

 [Dependencies::Stats]
 - Dependencies changed since 1.000001, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 (suggests: ↑1)
 - test: -4 (recommends: ↑1 -2)

 [Misc]
 - Improve source side POD to reduce build side whitespace.

 [Tests]
 - update test::reportprereqs

1.000001 2014-07-18T15:14:08Z
 [00 Trivial]
 - Minor packaging cleanups.

 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - configure: (recommends: ↑1)
 - develop: +3 ↓3 -3 (suggests: +2 -1)
 - test: +2 ↓1 (recommends: +1 ↑1)

1.000000 2014-04-30T22:34:56Z
 [00 Minor]
 - No major code changes, but tooling changes are sufficient that downstream vendors may want to pay attention.

 [01 Tooling Change]
 - EUMM is now used instead of Module::Build
 - "Latest" Toolchain deps are now only recommended instead of mandatory

 [02 Version Scheme change]
 - Moving from 0.x.y to 1.xxxxxx
 - This is the primary reason for the 1.x on the box.

 [Deoendencies::Aggregate]
 - 5.008 : ? → 204
 - 5.010 : 180 → 183
 - 5.018 : ? → 170
 - 5.019 : 166 → 164

 [Dependencies::Stats]
 - Dependencies changed since 0.1.6, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +7 ↑2 -9
 - runtime: +1 ↓1
 - test: +5 ↓1 -2 (recommends: +4)

0.1.6 2013-11-29T14:52:26Z
 [Dependencies::Noteworthy]
 - use Module::Build 0.4203

 [Dependencies::Stats]
 - Dependencies changed since 0.1.5, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: ↑1

 [Documentation]
 - Reference Prereqs::MatchInstalled::All ( Issue #2 )

0.1.5 2013-11-22T08:56:22Z
 [00 Maintenance]
 - Changelog simplified with new Develop Changes stuff.

 [Dependences]
 - Upgrade Module::Build and Test::More

 [Dependencies::Stats]
 - Dependencies changed since 0.1.4, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +43 ↑2 -1 (recommends: -1, suggests: +1 -1)
 - test: ↑1 -3

 [Packaging]
 - Dzil 5 Build.PL prereqs merge implementation.

 [Tests]
 - Use Test::Compile::PerFile
 - Reportversions updated = no reporting develop deps

0.1.4 2013-09-25T22:27:11Z
 [Bugfix]
 - Always call _version_emulate, which forces Module::MetaData to determine $VERSION

0.1.3 2013-09-25T10:32:16Z
 [Bugfix]
 - Bump to get Test::Compile that isn't broken on Win32

 [Dependencies]
 - Whole bunch of new dependencies to support new test

 [Dependencies::Stats]
 - Dependencies changed since 0.1.2, see misc/*.deps* for details
 - develop: ↑1
 - test: +7

 [Tests]
 - Now has a basic functionality test

0.1.2 2013-08-31T02:47:34Z
 [!Minor]
 - Bump to get newer Test::Compile test.

 [Dependencies::Stats]
 - Dependencies changed since 0.1.1, see misc/*.deps* for details
 - develop: +1 ↑2 (suggests: ↑1)
 - test: +2 -2

0.1.1 2013-08-07T22:34:28Z
 [Dependencies]
 - Need Capture Tiny for Test::Compile
 - Upgrade Module::Build

 [Dependencies::Stats]
 - Dependencies changed since 0.1.0, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: ↑1 (suggests: ↑1)
 - test: +2 -2

 [Internals]
 - Reimplement user module satisfaction code as a method so its easier to override in a child class.

0.1.0 2013-07-31T15:59:15Z
 - First version. Dogfood friendly, but no tests.


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.012.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4204"
  },
  "configure_requires" => {
    "Module::Build" => "0.4204"
  },
  "dist_abstract" => "Depend on versions of modules the same as you have installed",
  "dist_author" => [
    "Kent Fredric <kentfredric\@gmail.com>"
  ],
  "dist_name" => "Dist-Zilla-Plugin-Prereqs-MatchInstalled",
  "dist_version" => "0.1.7",
  "license" => "perl",
  "module_name" => "Dist::Zilla::Plugin::Prereqs::MatchInstalled",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Dist::Zilla::Role::PrereqSource" => 0,
    "Module::Data" => 0,
    "Moose" => 0,
    "MooseX::Types::Moose" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Cwd" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Copy::Recursive" => 0,
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "Path::FindDev" => 0,
    "Path::Tiny" => 0,
    "Test::DZil" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "1.001002",
    "version" => "0.9907"
  }
);


my %fallback_build_requires = (
  "Cwd" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Copy::Recursive" => 0,
  "File::Spec::Functions" => 0,
  "List::Util" => 0,
  "Module::Build" => "0.4204",
  "Path::FindDev" => 0,
  "Path::Tiny" => 0,
  "Test::DZil" => 0,
  "Test::Fatal" => 0,
  "Test::More" => "1.001002",
  "version" => "0.9907"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.008.
use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Depend on versions of modules the same as you have installed",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Dist-Zilla-Plugin-Prereqs-MatchInstalled",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008",
  "NAME" => "Dist::Zilla::Plugin::Prereqs::MatchInstalled",
  "PREREQ_PM" => {
    "Dist::Zilla::Role::PrereqSource" => 0,
    "Dist::Zilla::Util::ConfigDumper" => 0,
    "Module::Data" => 0,
    "Moose" => 0,
    "MooseX::Types::Moose" => 0,
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dist::Zilla::Util::Test::KENTNL" => "1.003001",
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Module::Metadata" => 0,
    "Test::DZil" => 0,
    "Test::Fatal" => 0,
    "Test::More" => 0
  },
  "VERSION" => "1.001004",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Dist::Zilla::Role::PrereqSource" => 0,
  "Dist::Zilla::Util::ConfigDumper" => 0,
  "Dist::Zilla::Util::Test::KENTNL" => "1.003001",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Module::Data" => 0,
  "Module::Metadata" => 0,
  "Moose" => 0,
  "MooseX::Types::Moose" => 0,
  "Test::DZil" => 0,
  "Test::Fatal" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
